zerver/tests/test_scheduled_messages.py

```diff
@@ -81,6 +89,203 @@ def test_schedule_message(self) -> None:
+        logger = mock.Mock()
+        # No scheduled message
+        self.assertFalse(try_deliver_one_scheduled_message(logger))
+
+        content = "Test message"
+        scheduled_delivery_datetime = timezone_now() + datetime.timedelta(minutes=5)
+        scheduled_delivery_timestamp = int(scheduled_delivery_datetime.timestamp())
+        othello = self.example_user("othello")
+        response = self.do_schedule_message(
+            "direct", [othello.email], content + " 3", scheduled_delivery_timestamp
+        )
+        self.assert_json_success(response)
+        scheduled_message = self.last_scheduled_message()
+
+        # mock current time to be greater than the scheduled time.
+        more_than_scheduled_delivery_datetime = scheduled_delivery_datetime + datetime.timedelta(
+            minutes=1
+        )
+        with mock.patch(
+            "zerver.actions.scheduled_messages.timezone_now",
+            return_value=more_than_scheduled_delivery_datetime,
+        ):
+            with mock.patch(
+                "zerver.actions.message_send.timezone_now",
+                return_value=more_than_scheduled_delivery_datetime,
+            ):
+                result = try_deliver_one_scheduled_message(logger)
+                self.assertTrue(result)
+                logger.info.assert_called_once_with(
+                    "Sending scheduled message %s with date %s (sender: %s)",
+                    scheduled_message.id,
+                    scheduled_message.scheduled_timestamp,
+                    scheduled_message.sender_id,
+                )
+                scheduled_message.refresh_from_db()
+                assert isinstance(scheduled_message.delivered_message_id, int)
+                self.assertEqual(scheduled_message.delivered, True)
+                self.assertEqual(scheduled_message.failed, False)
+                delivered_message = Message.objects.get(id=scheduled_message.delivered_message_id)
+                self.assertEqual(delivered_message.content, scheduled_message.content)
+                self.assertEqual(
+                    delivered_message.rendered_content, scheduled_message.rendered_content
+                )
+                self.assertEqual(delivered_message.date_sent, more_than_scheduled_delivery_datetime)
+
+    def verify_deliver_scheduled_message_failure(
+        self, scheduled_message: ScheduledMessage, logger: mock.Mock
+    ) -> None:
+        result = try_deliver_one_scheduled_message(logger)
+        self.assertTrue(result)
+        scheduled_message.refresh_from_db()
+        calls = [
+            mock.call(
+                "Sending scheduled message %s with date %s (sender: %s)",
+                scheduled_message.id,
+                scheduled_message.scheduled_timestamp,
+                scheduled_message.sender_id,
+            ),
+            mock.call("Failed with message: %s", scheduled_message.failure_message),
```

Comment:
It'd be nice if these tests passed the expected failure message string to `verify_deliver_scheduled_message_failure` so we can confirm they failed for the right reason; otherwise this could mask a bug where all these failure tests fail due to the same test setup bug.

Reply:
Done.